### PR TITLE
Merge IndexCoord and DataCoord

### PR DESCRIPTION
**Pull request by xiaocai2333**
Signed-off-by: cai.zhang <cai.zhang@zilliz.com>
issue: #21269

### CODE REVIEW

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/datacoord/compaction_test.go [hunk 1]
```diff
@@ -322,18 +322,18 @@ func TestCompactionPlanHandler_handleMergeCompactionResult(t *testing.T) {
 		plans:    plans,
 		sessions: sessions,
 		meta:     meta,
-		segRefer: &SegmentReferenceManager{
-			segmentsLock: map[UniqueID]map[UniqueID]*datapb.SegmentReferenceLock{},
-		},
+		//segRefer: &SegmentReferenceManager{
```
Comment (czs007) [inline internal/datacoord/compaction_test.go:325]: just delete this.

File: internal/rootcoord/root_coord.go [hunk 2]
```diff
@@ -274,16 +273,16 @@ func (c *Core) SetDataCoord(ctx context.Context, s types.DataCoord) error {
...
-		return err
-	}
-	if err := s.Start(); err != nil {
-		return err
-	}
-	c.indexCoord = s
-	return nil
-}
+//func (c *Core) SetIndexCoord(s types.IndexCoord) error {
+//	if err := s.Init(); err != nil {
```
Comment (czs007) [inline internal/rootcoord/root_coord.go:277]: just delete this.

File: internal/indexcoord/index_coord.go [hunk 3]
```diff
@@ -1134,23 +1135,23 @@ func (i *IndexCoord) tryAcquireSegmentReferLock(ctx context.Context, buildID Uni
 	// IndexCoord use buildID instead of taskID.
 	log.Info("try to acquire segment reference lock", zap.Int64("buildID", buildID),
 		zap.Int64("nodeID", nodeID), zap.Int64s("segIDs", segIDs))
-	ctx1, cancel := context.WithTimeout(ctx, reqTimeoutInterval)
```
Comment (czs007) [inline internal/indexcoord/index_coord.go:1137]: just delete this

File: internal/indexcoord/index_builder.go [hunk 4]
```diff
@@ -478,13 +478,13 @@ func (ib *indexBuilder) releaseLockAndResetNode(buildID UniqueID, nodeID UniqueI
 }
 
 func (ib *indexBuilder) releaseLockAndResetTask(buildID UniqueID, nodeID UniqueID) error {
-	if nodeID != 0 {
```
Comment (czs007) [inline internal/indexcoord/index_builder.go:481]: rename this function

File: internal/indexcoord/index_builder.go [hunk 5]
```diff
@@ -462,12 +462,12 @@ func (ib *indexBuilder) dropIndexTask(buildID, nodeID UniqueID) bool {
 }
 
 func (ib *indexBuilder) releaseLockAndResetNode(buildID UniqueID, nodeID UniqueID) error {
```
Comment (czs007) [inline internal/indexcoord/index_builder.go:464]: rename this function

File: internal/datacoord/meta_test.go [hunk 6]
```diff
@@ -613,42 +629,42 @@ func TestUpdateFlushSegmentsInfo(t *testing.T) {
...
-
-	keys, _, err = kvClient.LoadWithPrefix(util.FlushedSegmentPrefix)
-	assert.Nil(t, err)
-	assert.Equal(t, 1, len(keys))
-
-	segmentID, err := strconv.ParseInt(filepath.Base(keys[0]), 10, 64)
-	assert.Nil(t, err)
-	assert.Equal(t, 100, int(segmentID))
-}
+//func TestSaveHandoffMeta(t *testing.T) {
```
Comment (czs007) [inline internal/datacoord/meta_test.go:632]: delete this

File: internal/indexcoord/index_builder.go (hunk 5 above)
Comment (xiaocai2333) [reply internal/indexcoord/index_builder.go:464]: it will be removed.

**Comment by xiaocai2333** [comment]
rerun ut

File: configs/milvus.yaml [hunk 7]
```diff
@@ -307,6 +308,13 @@ dataCoord:
     missingTolerance: 86400 # file meta missing tolerance duration in seconds, 60*24
     dropTolerance: 86400 # file belongs to dropped entity tolerance duration in seconds, 60*24
 
+  bindIndexNodeMode:
```
Comment (czs007) [inline configs/milvus.yaml:311]: delete bindIndexNodeMode from indexCoord

File: internal/datacoord/handler.go [hunk 8]
```diff
@@ -148,13 +149,76 @@ func (h *ServerHandler) GetQueryVChanPositions(channel *channel, partitionID Uni
 		}
 	}
 
+	//var (
```
Comment (czs007) [inline internal/datacoord/handler.go:152]: just delete this

File: internal/datacoord/handler.go [hunk 9]
```diff
@@ -148,13 +149,76 @@ func (h *ServerHandler) GetQueryVChanPositions(channel *channel, partitionID Uni
...
+	//	UnflushedSegmentIds: growingIDs.Collect(),
+	//	DroppedSegmentIds:   droppedIDs.Collect(),
+	//}
+
 	return &datapb.VchannelInfo{
 		CollectionID:        channel.CollectionID,
 		ChannelName:         channel.Name,
 		SeekPosition:        h.GetChannelSeekPosition(channel, partitionID),
 		FlushedSegmentIds:   indexedIDs.Collect(),
 		UnflushedSegmentIds: unIndexedIDs.Collect(),
```
Comment (czs007) [inline internal/datacoord/handler.go:219]: In the future, suggest change the name:
FlushedSegmentIds => IndexedSegmentIDs
UnflushedSegmentIds => unINdexedIDs

File: internal/metastore/kv/datacoord/kv_catalog.go [hunk 10]
```diff
@@ -171,15 +172,15 @@ func (kc *Catalog) AlterSegment(ctx context.Context, newSegment *datapb.SegmentI
 		return err
 	}
 	maps.Copy(kvs, segmentKvs)
-	if newSegment.State == commonpb.SegmentState_Flushed && oldSegment.State != commonpb.SegmentState_Flushed {
```
Comment (czs007) [inline internal/metastore/kv/datacoord/kv_catalog.go:174]: why comment this ?

File: internal/metastore/kv/datacoord/kv_catalog.go [hunk 11]
```diff
@@ -247,16 +248,16 @@ func (kc *Catalog) AlterSegmentsAndAddNewSegment(ctx context.Context, segments [
...
-			// should be a faked segment, we create flush path directly here
-			flushSegKey := buildFlushedSegmentPath(newSegment.GetCollectionID(), newSegment.GetPartitionID(), newSegment.GetID())
-			clonedSegment := proto.Clone(newSegment).(*datapb.SegmentInfo)
-			clonedSegment.IsFake = true
-			segBytes, err := marshalSegmentInfo(clonedSegment)
-			if err != nil {
-				return err
-			}
-			kvs[flushSegKey] = segBytes
+			//} else {
```
Comment (czs007) [inline internal/metastore/kv/datacoord/kv_catalog.go:251]: what if GetNumOfRows ==0 ?

File: internal/metastore/kv/indexcoord/kv_catalog.go [hunk 12]
```diff
@@ -60,7 +60,19 @@ func (kc *Catalog) ListIndexes(ctx context.Context) ([]*model.Index, error) {
 			return nil, err
 		}
 
-		indexes = append(indexes, model.UnmarshalIndexModel(meta))
+		indexes = append(indexes, &model.Index{
```
Comment (czs007) [inline internal/metastore/kv/indexcoord/kv_catalog.go:63]: Is this file still necessary to exist?

File: internal/datacoord/server.go [hunk 13]
```diff
@@ -419,30 +434,29 @@ func (s *Server) initServiceDiscovery() error {
...
+				return err
+			}
+		}
 	}
-	s.qcEventCh = s.session.WatchServices(typeutil.QueryCoordRole, qcRevision+1, nil)
+	s.inEventCh = s.session.WatchServices(typeutil.IndexNodeRole, inRevision+1, nil)
 
-	s.segReferManager, err = NewSegmentReferenceManager(s.kvClient, serverIDs)
-	return err
+	return s.clearLegacyDataFromOldVersion()
```
Comment (czs007) [inline internal/datacoord/server.go:459]: this should be optional

File: internal/metastore/kv/datacoord/kv_catalog.go (hunk 11 above)
Comment (xiaocai2333) [reply internal/metastore/kv/datacoord/kv_catalog.go:251]: will reset it.

File: internal/metastore/kv/indexcoord/kv_catalog.go (hunk 12 above)
Comment (xiaocai2333) [reply internal/metastore/kv/indexcoord/kv_catalog.go:63]: it is unused, just for make

File: internal/datacoord/server.go (hunk 13 above)
Comment (xiaocai2333) [reply internal/datacoord/server.go:459]: ok, will log it instead of return error

Comment (czs007) [reply internal/datacoord/server.go:459]: just remove this

File: configs/milvus.yaml (hunk 7 above)
Comment (xiaocai2333) [reply configs/milvus.yaml:311]: will delete it next pr.

File: internal/metastore/kv/datacoord/kv_catalog.go (hunk 11 above)
Comment (xiaocai2333) [reply internal/metastore/kv/datacoord/kv_catalog.go:251]: it is used to write flushed segment, it is no longer necessary.

File: internal/datacoord/server.go (hunk 13 above)
Comment (xiaocai2333) [reply internal/datacoord/server.go]: ok, remove it

File: internal/datacoord/handler.go (hunk 9 above)
Comment (xiaocai2333) [reply internal/datacoord/handler.go:219]: change it in next pr.

File: internal/datacoord/handler.go (hunk 8 above)
Comment (xiaocai2333) [reply internal/datacoord/handler.go:152]: it used to statistics the indexed segments, flushed segments and growing segments. Need QueryCoord to support this.
